services/scheduler/src/main/java/org/mule/service/scheduler/internal/DefaultScheduler.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.service.scheduler.internal;
+
+import static java.lang.System.lineSeparator;
+import static java.lang.System.nanoTime;
+import static java.lang.Thread.currentThread;
+import static java.util.Collections.synchronizedSet;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import org.mule.runtime.core.api.scheduler.Scheduler;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.RunnableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Proxy for a {@link ScheduledExecutorService} that adds tracking of the source of the dispatched tasks.
+ *
+ * @since 4.0
+ */
+class DefaultScheduler extends AbstractExecutorService implements Scheduler {
+
+  /**
+   * Forced shutdown delay. The time to wait while threads are being interrupted.
+   */
+  private static final long FORCEFUL_SHUTDOWN_TIMEOUT_SECS = 5;
+
+  private static final Logger logger = LoggerFactory.getLogger(DefaultScheduler.class);
+
+  private final ExecutorService executor;
+  private final ScheduledExecutorService scheduledExecutor;
+
+  /**
+   * Wait condition to support awaitTermination
+   */
+  private final CountDownLatch terminationLatch = new CountDownLatch(1);
+
+  private Set<RunnableFutureDecorator<?>> currentTasks = synchronizedSet(new HashSet<>());
```

Comment:
I was able to replace it with a ConcurrentHashMap, that should improve performance.